src/VisualStudio/Core/Def/Storage/AbstractCloudCacheService.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Storage;
+using Microsoft.VisualStudio.RpcContracts.Caching;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Storage
+{
+    /// <summary>
+    /// Base implementation of <see cref="IRoslynCloudCacheService"/> that sits on top of the VS platform's <see
+    /// cref="ICacheService"/>, taking care of all common logic.
+    /// </summary>
+    internal abstract class AbstractCloudCacheService : IRoslynCloudCacheService
+    {
+        protected readonly ICacheService CacheService;
+
+        protected AbstractCloudCacheService(ICacheService cacheService)
+        {
+            CacheService = cacheService;
+        }
+
+        private static CacheItemKey Convert(RoslynCloudCacheItemKey key)
+            => new(Convert(key.ContainerKey), key.ItemName) { Version = key.Version };
+
+        private static CacheContainerKey Convert(RoslynCloudCacheContainerKey containerKey)
+            => new(containerKey.Component, containerKey.Dimensions);
+
+        public abstract void Dispose();
+
+        public ValueTask DisposeAsync()
```

Comment:
😕 Why is `Dispose()` abstract but `DisposeAsync()` is sealed?

Reply:
yup.  i don't love this, but the general idea (as i understand it) is trhat you're supposed to be both IAsyncDisposable and Disposable.  Building the former on the latter is trivial.  HOwever, determining how each service shoudl be synchronously disposable is up to that service itself.